Get rid of BasePath on IConfigurationBuilder

kirthik opened the pull request:
cc @divega

divega commented on src/Microsoft.Framework.Configuration.Abstractions/IConfigurationBuilder.cs:
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Framework.Configuration
     public interface IConfigurationBuilder
     {
-        string BasePath { get; }
+        Dictionary<string, object> Properties { get; set; }
```
Can we remove `set`?

divega commented on src/Microsoft.Framework.Configuration/ConfigurationBuilder.cs:
```diff
@@ -24,7 +25,7 @@ public ConfigurationBuilder(string basePath, params IConfigurationSource[] sourc
             }
 
-            BasePath = basePath;
+            _properties["BasePath"] = string.Empty;
```
We shouldn't need to do this, i.e. ConfigurationBuilder shouldn't need to know about any specific properties.

kirthik replied:
How would a case like this work then - 
```
var builder = new ConfigurationBuilder();
builder.AddIniFile(_iniConfigFilePath);
```

divega replied:
See my other comment below about moving the path resolution logic to live along side the `SetPath()` method. This case should be fine because the path resolution logic can default to whatever it needs to (`string.Empty`?) if the property is not set.

divega replied:
BTW, the `basePath` argument in this constructor needs to be removed.

divega commented on src/Microsoft.Framework.Configuration/Helper/ConfigurationHelper.cs:
```diff
@@ -10,19 +10,11 @@ public static class ConfigurationHelper
     {
         public static string ResolveConfigurationFilePath(IConfigurationBuilder configuration, string path)
         {
+            string basePath = configuration.Properties["BasePath"].ToString();
```
Configuration "core" should become ignorant of any properties, including BasePath. Any common logic to get the base path or to resolve it should move to the new file-based package.

davidfowl commented:
Can we avoid the .File namespace?

kirthik commented:
I am not using .File namespace in classes. Do you mean the Package name?

divega commented:
@davidfowl naming is all tentative in thus PR. We actually were about to start asking for feedback on it, for the package and the extension method class as well as for the dictionary property on `ConfigurationBuilder`.

Cc @Eilon @lodejard

Eilon commented:
Would be happy to help with names. But is there any reason not to put the BasePath extensions in the main config package?

kirthik commented:
cc @divega

divega commented:
> Would be happy to help with names. But is there any reason not to put the BasePath extensions in the main config package?

The BCL team wanted the core of Configuration not to be coupled to file system concepts. Sources can be specific to scenarios in which there is a file system, as our file-based sources are.

Eilon commented:
Yeah I remember that, but in some sense this doesn't really couple any APIs. It's more of an optional extension method that is commonly used. It's a fine line, I agree, but having a whole extra package for this one little feature seems excessive, no?

divega commented:
@Eilon the `SetBasePath()` extension method doesn't do anything but adding a property to the bag, but all the file-based sources will now call another extension method defined on the configuration builder to do the `Path.Combine()` logic, and that lives in the new package now. 

I guess we could copy and paste that logic into all our file-based sources but that seems wrong. The new package will be common dependency to all file-based sources. We haven't looked around maybe there is more common stuff that could live there, so I don't mind.

divega commented on src/Microsoft.Framework.Configuration.Abstractions/IConfigurationBuilder.cs:
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Framework.Configuration
     public interface IConfigurationBuilder
     {
-        string BasePath { get; }
+        Dictionary<string, object> Properties { get; }
```
@Eilon does the name `Properties` seem ok for you for the property bag on `ConfigurationBuilder`? Other alternatives could be:

`ExtendedProperties`
`BuildTimeProperties` :trollface:

Eilon replied:
I'm a fan of just `Properties` - that's a pretty common name used in lots of places for "a bag of stuff."

divega replied:
:+1:

divega commented on src/Microsoft.Framework.Configuration.File/Helper/ConfigurationHelper.cs:
```diff
@@ -10,18 +10,13 @@ public static class ConfigurationHelper
     {
         public static string ResolveConfigurationFilePath(IConfigurationBuilder configuration, string path)
```
I think we should move this to FileConfigurationBuilderExtensions and make it an extension method, e.g. `ResolveFilePath(this IConfigurationBuilder configurationBuilder, string filePath)`.

divega replied:
@Eilon please help name the extension method. Proposal above.

Eilon replied:
`GetConfigFilePath(this IConfigurationBuilder configuration, string path)`?

divega replied:
I am ok with this, although I am not sure about abbreviating `Configuration` (I am not sure we have so far done it) and I still would prefer to have `Resolve` in the name.

divega commented on src/Microsoft.Framework.Configuration.File/FileConfigurationbuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+                throw new ArgumentNullException(nameof(basePath));
+            }
+
+            if (builder.Properties.ContainsKey("BasePath"))
```
Just use `builder.Properties["BasePath"] = basePath;` regardless.

divega commented on src/Microsoft.Framework.Configuration.File/Helper/ConfigurationHelper.cs:
```diff
@@ -10,18 +10,13 @@ public static class ConfigurationHelper
-                        nameof(configuration.BasePath)));
-                }
-                else
+                string basePath = configuration.Properties["BasePath"].ToString();
```
I think we should just do a `(string)` cast. If it fails then the wrong data got into the property somehow.

lodejard replied:
Call TryGet instead of ContainsKey, then `(string)` case like @divega mentioned

divega commented on src/Microsoft.Framework.Configuration/ConfigurationBuilder.cs:
```diff
@@ -35,17 +29,20 @@ public IEnumerable<IConfigurationSource> Sources
         }
 
-        public string BasePath
+        public Dictionary<string, object> Properties
```
You could just do an auto-property: 
``` C#
public Dictionary<string, object> Properties { get;  } = new Dictionary<string, object>();
```

Eilon commented:
@divega if it's brought in transitively I guess it's not as bad from a user experience perspective. But we also need to be a bit careful of package explosion.

divega commented:
> But we also need to be a bit careful of package explosion.

Agreed, I just think in this case it might be ok.

divega commented on Configuration.sln:
```diff
@@ -41,6 +41,10 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Framework.Configu
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Framework.Configuration.Binder.Test", "test\Microsoft.Framework.Configuration.Binder.Test\Microsoft.Framework.Configuration.Binder.Test.xproj", "{AE6FFE9B-6378-4D57-AA24-7D257F18B235}"
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Framework.Configuration.File", "src\Microsoft.Framework.Configuration.File\Microsoft.Framework.Configuration.File.xproj", "{881E7CBC-492C-47C5-98A6-61DD1C753EE6}"
```
@Eilon please help name the package name. `Microsoft.Framework.Configuration.File`  was the simplest thing we came up with, but not necessary great.

Eilon replied:
Hmm, a name for a package with one type that has 2 or 3 methods in it... :smile: 

`Microsoft.Framework.Configuration.FileHelpers` ? Kind of a lame name, but that's all these are. None are absolutely required to be used (anyone can easily write the code for what they do).

divega replied:
I don't love `FileHelpers` but I can live with it.

divega replied:
I am liking `Microsoft.Framework.Configuration.FileExtensions` more and more for reasons explained in other comments. @Eilon would you be ok with that?

Eilon replied:
Sure.

divega commented on src/Microsoft.Framework.Configuration.File/FileConfigurationbuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Microsoft.Framework.Configuration
+{
+    public static class FileConfigurationBuilderExtensions
```
@Eilon you can as well help name the sponsor class (hmmm, I think I am spamming you already :smile:)

Eilon replied:
Given my (lame) proposed package name, the matching type name would be `FileHelpersConfigurationBuilderExtensions`.

divega replied:
I looked at the sources and we consistently use just "XyzConfigurationExtensions"  although the target is `IConfigurationBuilder`. I think we should do the same here.

kirthik replied:
So should it be `FileHelpersConfigurationExtensions`?

divega replied:
Hmmm, I used to think so, but having both helpers and extensions at the same time in the name  sounds goofy. Maybe the class should be called `FileHelpers` or maybe the package should be called `FileExtensions`... We can wait until we discuss this with Fowler tomorrow, in case we decide to take a completely different approach.

divega replied:
I guess the name of the class can be `FileConfigurationExtensions` then.

Eilon commented on src/Microsoft.Framework.Configuration.File/FileConfigurationbuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="basePath">Absolute path of file based providers.
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder SetBasePath(this IConfigurationBuilder builder, string basePath)
```
Should there be a `GetBasePath()` companion to this? Trivial to implement, and appears more symmetrical (rare to get a set without a get). I realize our own code won't need this, but it might be nice for other consumers that might just want the path to do their own manipulation (rather than use the resolve helper method).

divega replied:
Sounds good to me. We can add it.

lodejard commented on src/Microsoft.Framework.Configuration.File/Helper/ConfigurationHelper.cs:
```diff
@@ -10,18 +10,13 @@ public static class ConfigurationHelper
-                else
+                string basePath = configuration.Properties["BasePath"].ToString();
+
+                if (!Path.IsPathRooted(path))
```
Don't call IsPathRooted - calling Path.Combine all the time should return the right-hand path only if it is rooted.

davidfowl commented:
This whole change makes me wonder if using the static appbase is a lesser evil....

divega commented:
Would that couple Configuration more with ASP.NET?

davidfowl commented:
@divega typo -> DNX not ASP.NET. Also we'd cross compile and do the right thing to get the Application path on each supported platform. We were trying to get away from that but this is getting uglier by the minute :smile:

divega commented:
@davidfowl I asked because I honestly didn't know where the static appbase thing would come from :)

davidfowl commented:
@divega I actually know how we can do it without coupling to the DNX at all.

divega commented:
Regardless, what is the part that looks ugly to you? Is it the new package?

The usage is pretty simple:

``` C#
var builder = new ConfigurationBuilder()
 .SetBasePath(basePath) 
 .AddIniFile("config.ini")
 .AddJsonFile("config.json");
```

The mechanism could be used to inject other things for Configuration sources, e.g. it was suggested that it would work to address #183.

davidfowl commented:
Calling SetBasePath is ugly. It was already ugly that we passed it in the ctor but this is much worse from an aesthetic POV. I think having it be an option is fine but maybe we need to support relative paths by default again.

divega commented:
OK,will you have time tomorrow? We need to decide what we do for beta8.

divega commented:
We talked to @davidfowl offline about this. The conclusion is that we will make setting the base path (with `SetBasePath()`) optional (you won't need to do it by default, but it might be useful for test scenarios or any time you want to artificially set it). The logic will be in the new `Microsoft.Framework.Configuration.FileExtensions` (aka `FileHelppers`) package in the ResolveConfigurationFilePath method, and it will require cross compilation because there is no uniform way to get the right path across all platforms supported by Configuration. Here is the summary of how it will work based on @davidfowl's analysis :
* __.NET Core (Contract based), this includes dnxcore50__ - AppContext.BaseDirectory (this also works on .net 4.6 but we’ll ignore that)
* __.NET 4.x - AppDomain.CurrentDomain.BaseDirectory__
* __DNX4.x based__ - (string)AppDomain.CurrentDomain.GetData(“APP_CONTEXT_BASE_DIRECTORY”)

davidfowl commented:
Instead of cross compiling for DNX specifically you can make the .NET 4.x version check `(string)AppDomain.CurrentDomain.GetData("APP_CONTEXT_BASE_DIRECTORY")` before looking at the base directory.

divega commented:
@kirthik please let me know if you can handle this as part of this PR. I think it would also be reasonable to file a bug and do it in RC (unless others in the thread pushback).

divega commented:
I went ahead and filed #295 to make `SetBasePath()` optional. I think we should still do it (for RC) maybe but we should not block this PR on that.

kirthik commented:
@divega : I addressed comments around assembly names and other PR comments in existing code. I will address #295 in a separate PR(in RC may be) as you mentioned above.

divega commented on src/Microsoft.Framework.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -0,0 +1,70 @@
+            return null;
+        }
+
+        public static string GetConfigurationFilePath(this IConfigurationBuilder configuration, string path)
```
Reuse `GetBasePath()` rather than duplicating the logic.

divega replied:
The name of the IConfigurationBuilder parameter should be descriptive an consistent (I filed a bug last night for the rest of the Configuration code base). IMO 'configuration' is not ok. I think `configurationBuilder` is the best, but 'builder' (as we are using in the method above) maybe good enough.

divega replied:
Moreover, it is now clear that this method is just plain vanilla `Path.Combine(configurationBuilder.GetBasePath(), filePath)`. I am thinking we don't really need it anymore. Even the logic for #295 will go inside `GetBasePath()`. Can we get rid of it?

kirthik replied:
Makes sense. I will get right of `GetConfigurationFilePath` and will add Path.Combine in `AddxyzExtensions.cs`

divega commented on src/Microsoft.Framework.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -0,0 +1,70 @@
+                return (string)basePath;
+            }
+
+            return null;
```
I suspect string.Empty is a more useful default to return here, e.g. `Path.Combine()` throws for null. Thoughts?

kirthik replied:
makes sense specially now since we are going to remove `GetConfigurationFilePath` extension

divega commented on test/Microsoft.Framework.Configuration.Test/ConfigurationTest.cs:
```diff
@@ -12,20 +12,13 @@ namespace Microsoft.Framework.Configuration.Test
     {
         [Fact]
-        public void SetBasePathThroughConstructor()
+        public void SetBasePathThroughBuilder()
```
Can you move this test to `Microsoft.Framework.Configuration.FileExtensions.Test`?

divega replied:
Or otherwise avoid creating that new package and have all the unit tests here instead? Not sure if there is a hard rule about that.

cc @Eilon

Eilon replied:
The goal is to have one unit test assembly per each runtime assembly.

kirthik replied:
I will move it to `Microsoft.Framework.Configuration.FileExtensions.Test`

divega commented on test/Microsoft.Framework.Configuration.Test/ConfigurationHelperTest.cs:
```diff
@@ -17,29 +16,12 @@ public void ResolveFilePath()
-            {
-                ConfigurationHelper.ResolveConfigurationFilePath(mockSourceRoot, testFile);
-            });
+            mockSourceRoot.SetBasePath(testDir);
```
Same consideration for this test. Either it belong in `Microsoft.Framework.Configuration.FileExtensions.Test` or in `Microsoft.Framework.Configuration.Test` but in a class that is for tests for FileConfigurationExtensions class.

kirthik replied:
Will move this to `FileConfigurationExtensionsTest` class

divega commented on src/Microsoft.Framework.Configuration.Xml/XmlConfigurationExtensions.cs:
```diff
@@ -56,7 +55,7 @@ public static IConfigurationBuilder AddXmlFile(this IConfigurationBuilder config
             }
 
-            var fullPath = ConfigurationHelper.ResolveConfigurationFilePath(configuration, path);
+            var fullPath = configuration.GetConfigurationFilePath(path);
```
Consider fixing #296 while you are editing all the `AddXyz` methods.

kirthik replied:
Makes sense. Will do it.

kirthik commented:
@divega : updated

divega commented on src/Microsoft.Framework.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="basePath">Absolute path of file based providers.
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder SetBasePath(this IConfigurationBuilder builder, string basePath)
```
Nit: `this IConfigurationBuilder configurationBuilder`

divega commented on src/Microsoft.Framework.Configuration.FileExtensions/FileConfigurationExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+        /// </summary>
+        /// <param name="builder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static string GetBasePath(this IConfigurationBuilder builder)
```
Nit: `this IConfigurationBuilder configurationBuilder`

divega commented:
:shipit:

kirthik commented:
merged bf62a69de10a50f01ddc4b9049abe8572c4995d8
